content/blog/2025-06-06-a-plan-for-simd.md

```diff
@@ -0,0 +1,142 @@
+The main reason to go smaller is when the loop has a high probability of consuming less than 256 bits of input.
+
+In the other direction, the majority of shipping AVX-512 chips are double-pumped, so code written to use 512 bits is not significantly faster (I assert this based on some serious experimentation on a Zen 5 laptop).
+I also expect this state of affairs to continue for a while; AMD won this design war and Intel is struggling to catch up after flailing for many years.
+If and when a significant number of true 512 bit chips ship, that would justify more work to write 512-optimized variants.
+
+For map-like workloads, I propose adding pulp-like associated natural-width types and methods to the Simd trait.
+Following Highway, I think it also makes sense to have some more methods that work in 128 bit blocks.
+A motivating use case is f32 color space conversion, where the alpha channel is passed through unmodified; a very reasonable implementation strategy is to do the nonlinear conversion for all channels, then do a blend operation on lane 3 of 128 bit blocks.
+(Additional note: in experiments, trying to do this absolutely broke autovectorization).
+
+## Light use of macros
+
+I am quite concerned about the compile time.
+Cold build of fearless\_simd\#2 on an M1 Max is 3.25s.
+Considering that building out the full cross-product of sizes and types will cause about an order of magnitude increase in size, it’s clear that the library has the potential for major impact on compile times.
+
+Using \-Z self-profile to investigate, 87.6% of the time is in expand\_crate, which I believe is primarily macro expansion \[an expert in rustc can confirm or clarify\].
+This is not hugely surprising, as (following the example of pulp), declarative macros are used very heavily.
+A large fraction of that is the safe wrappers for intrinsics (corresponding to [core\_arch](https://docs.rs/pulp/latest/pulp/core_arch/x86/index.html) in pulp).
+
+I believe that using codegen to expand out the macros before crate publish time will help greatly with compile times, but this needs to be experimentally validated.
+A possible downside is the size of the crate (especially uncompressed), but I expect zlib compression to be very effective given the repetitive, boilerplate nature of the contents.
+
+One use of macros will remain: simd\_dispatch as a declarative macro to generate the dispatch wrappers.
+Likely the proposed [declarative macro improvements](https://rust-lang.github.io/rust-project-goals/2025h1/macro-improvements.html) could help a lot here.
+I’m especially positive about the ability to write attributes as a declarative macro, as that would reduce the stuttering in the existing syntax.
+
+## Topics discussed in the blog
+
+Some of these may be expanded, but I’ll list them as they’re relevant:
+
+* dispatch done by passing around \`Level\` enum; minimize cost of detection  
```

Comment:
... around a `Level` ...

But also, this is the first mention here of a Level enum in this document. I wish for Tufte style marginalia.